File: third-person/public/camera.js
class Camera {
    constructor(thing, radius, rx, ry) {
        this.thing = thing
        this.radius = radius
        this.x
        this.y
        this.z
        this.rx = rx
        this.ry = ry
        this.update()
    }
    update(interpolation) {
        let sin_x = Math.sin(this.rx)
        let cos_x = Math.cos(this.rx)
        let sin_y = Math.sin(this.ry)
        let cos_y = Math.cos(this.ry)

        let vx = this.thing.OX + interpolation * (this.thing.X - this.thing.OX)
        let vy = this.thing.OY + interpolation * (this.thing.Y - this.thing.OY)
        let vz = this.thing.OZ + interpolation * (this.thing.Z - this.thing.OZ)

        this.x = vx - this.radius * cos_x * sin_y
        this.y = vy + this.radius * sin_x
        this.z = vz + this.radius * cos_x * cos_y

        this.y += this.thing.Height
    }
}

File: source/game/camera.c
#include "camera.h"

camera *camera_init(float radius) {
    camera *c = safe_calloc(1, sizeof(camera));
    c->radius = radius;
    return c;
}

void camera_update(camera *self) {
    float sin_x = sinf(self->rx);
    float cos_x = cosf(self->rx);
    float sin_y = sinf(self->ry);
    float cos_y = cosf(self->ry);

    float dx = -cos_x * sin_y;
    float dy = sin_x;
    float dz = cos_x * cos_y;

    self->x = self->target->x + self->radius * dx;
    self->y = self->target->z + self->radius * dy + self->target->height;
    self->z = self->target->y + self->radius * dz;
}
